sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,343 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class MethodsShouldUseBaseTypes : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3242";
+        private const string MessageFormat = "Use '{0}' here; it is a more general type than '{1}'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var violations = FindViolations(c).ToList();
+                violations.ForEach(c.ReportDiagnostic);
+            },
+            SyntaxKind.MethodDeclaration);
+        }
+
+        public IEnumerable<Diagnostic> FindViolations(SyntaxNodeAnalysisContext context)
+        {
+            var methodSymbol = context.SemanticModel.GetDeclaredSymbol(context.Node) as IMethodSymbol;
+
+            if (methodSymbol == null ||
+                methodSymbol.Parameters.Length == 0 ||
+                methodSymbol.IsOverride)
+            {
+                return Enumerable.Empty<Diagnostic>();
+            }
+
+            var parametersToCheck = new Dictionary<string, ParameterData>();
+            foreach (var parameter in methodSymbol.Parameters.Where(IsTrackedParameter))
+            {
+                parametersToCheck[parameter.Name] = new ParameterData(parameter);
+            }
+
+            var parameterUsesInMethod = context
+                .Node
+                .DescendantNodes()
+                .OfType<IdentifierNameSyntax>()
+                .Where(id => parametersToCheck.Values.Any(p => p.MatchesIdentifier(id, context.SemanticModel)))
+                .ToArray();
```

Comment:
I use ToList() in production code (when needed) and ToArray() while developing/in debugging. So if you ever see ToArray() in PR, it most likely means I forgot to remove it. Just like here :)